firedrake/output.py

```diff
@@ -651,3 +756,366 @@ def write(self, *functions, **kwargs):
+        nodes_per_entity, real_tensorproduct = sd_key
+        assert not real_tensorproduct
+        global_numbering = get_global_numbering(tV.mesh(), sd_key)
+        sectiondm = PETSc.DMShell().create(dm.comm)
+        sectiondm.setPointSF(dm.getPointSF())
+        sectiondm.setSection(global_numbering)
+    else:
+        sectiondm = tV._shared_data.node_set.halo.dm
+    sectiondm.setName(get_sectiondm_name(*sd_key))
+    return sectiondm
+
+
+class Dict(dict): pass
+
+
+class CheckpointFile(object):
+    _mesh_cache = weakref.WeakValueDictionary()
+
+    def __init__(self, filename, mode, comm=COMM_WORLD):
+        self.viewer = ViewerHDF5()
+        self.viewer.create(filename, mode=mode, comm=comm)
+        commkey = comm.py2f()
+        assert commkey != MPI.COMM_NULL.py2f()
+        self._meshes = self._mesh_cache.setdefault((filename, commkey), Dict())
+        self._function_spaces = {}
+        self._function_load_utils = {}
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        self.close()
+
+    def save_mesh(self, mesh, format=ViewerHDF5.Format.HDF5_PETSC):
+        dm = mesh.topology.topology_dm
+        if dm.getName() in self.h5pyfile.require_group("/topologies"):
+            return
+        if format not in [ViewerHDF5.Format.HDF5_PETSC, ViewerHDF5.Format.HDF5_XDMF]:
+            raise TypeError("Unknown output format %s." % format)
+        self.viewer.pushFormat(format=format)
+        if format == ViewerHDF5.Format.HDF5_PETSC:
+            dm.topologyView(viewer=self.viewer)
+            dm.coordinatesView(viewer=self.viewer)
+            dm.labelsView(viewer=self.viewer)
+        else:
+            dm.view(viewer=self.viewer)
+        self.viewer.popFormat()
+
+    def save_function_space(self, V):
+        if type(V.topological) is impl.MixedFunctionSpace:
+            base_path = os.path.join("/topologies",
+                                     PREFIX + "_mixed_function_spaces",
+                                     V.name)
+            self.h5pyfile.require_group(base_path)
+            self.set_attr(base_path, PREFIX + "_num_sub_spaces", V.num_sub_spaces())
+            for i, Vsub in enumerate(V):
+                path = os.path.join(base_path, str(i))
+                self.h5pyfile.require_group(path)
+                self.set_attr(path, PREFIX + "_mesh", Vsub.mesh().name)
+                self.set_attr(path, PREFIX + "_function_space", Vsub.name)
+                self.save_function_space(Vsub)
+            return
+        # Save mesh
+        self.save_mesh(V.mesh())
+        # Save function space
+        tV = V.topological
+        dm = tV.mesh().topology_dm
+        sectiondm = get_sectiondm_for_checkpointing(tV)
+        fs_dir_name = os.path.join("/topologies",
+                                   dm.getName(),
+                                   "dms",
+                                   sectiondm.getName(),
```

Comment:
`sectiondm` is internal, so we name it based on the shared data key, `(nodes_per_entity, real_tensorproduct)`, so that they never collide. It looks something like `sectiondm_1_3_3_False`.